From f71239ef54c368bfae8c29a9f312c7fc3f6091ed Mon Sep 17 00:00:00 2001
From: Konrad Dybcio <konrad.dybcio@oss.qualcomm.com>
Date: Fri, 3 Jul 2026 14:47:08 +0200
Subject: [PATCH 11/11] usb: misc: qcom_eud: Handle extra TCSR register

On some platforms, before EUD can be enabled, one must write a magic
bit in one of the spare registers of TCSR, otherwise EUD would seem to
enable correctly, yet it would never actually.

Handle that necessary write.

Assisted-by: Codex:gpt-5.5
Signed-off-by: Konrad Dybcio <konrad.dybcio@oss.qualcomm.com>
---
 drivers/usb/misc/qcom_eud.c | 28 ++++++++++++++++++++++++++++
 1 file changed, 28 insertions(+)

diff --git a/drivers/usb/misc/qcom_eud.c b/drivers/usb/misc/qcom_eud.c
index c3f46f30f304..367eddbc045c 100644
--- a/drivers/usb/misc/qcom_eud.c
+++ b/drivers/usb/misc/qcom_eud.c
@@ -11,6 +11,7 @@
 #include <linux/kernel.h>
 #include <linux/module.h>
 #include <linux/of.h>
+#include <linux/of_address.h>
 #include <linux/of_graph.h>
 #include <linux/phy/phy.h>
 #include <linux/platform_device.h>
@@ -491,10 +492,14 @@ static int eud_init_path(struct eud_chip *chip, struct device_node *np)
 static int eud_probe(struct platform_device *pdev)
 {
 	struct device_node *np = pdev->dev.of_node;
+	struct device_node *syscon_np;
 	struct eud_chip *chip;
 	struct resource *res;
 	int ret;
 
+	if (!qcom_scm_is_available())
+		return -EPROBE_DEFER;
+
 	chip = devm_kzalloc(&pdev->dev, sizeof(*chip), GFP_KERNEL);
 	if (!chip)
 		return -ENOMEM;
@@ -503,6 +508,29 @@ static int eud_probe(struct platform_device *pdev)
 
 	mutex_init(&chip->state_lock);
 
+	/* Some platforms first need to set a bit in a TCSR register to allow EUD enable */
+	syscon_np = of_parse_phandle(chip->dev->of_node, "qcom,tcsr-check", 0);
+	if (syscon_np) {
+		struct of_phandle_args args;
+		struct resource tcsr_res;
+
+		ret = of_address_to_resource(syscon_np, 0, &tcsr_res);
+		if (ret)
+			return ret;
+
+		of_node_put(syscon_np);
+
+		ret = of_parse_phandle_with_fixed_args(np, "qcom,tcsr-check", 2, 0, &args);
+		if (ret)
+			return ret;
+
+		of_node_put(args.np);
+
+		ret = qcom_scm_io_writel(tcsr_res.start + args.args[0], BIT(args.args[1]));
+		if (ret)
+			return dev_err_probe(chip->dev, ret, "Failed to set TCSR bit!\n");
+	}
+
 	for_each_child_of_node_scoped(np, child) {
 		ret = eud_init_path(chip, child);
 		if (ret)
-- 
2.55.0

